File: Res/Sorting/MergeSort/mergeSort.py
def mergeSort(list):
    n = len(list)
    # Base Case
    if n <= 1:
        return list
    
    mid = n // 2
    left_half = list[: mid]
    right_half = list[mid :]
    
    left_half = mergeSort(left_half)
    right_half = mergeSort(right_half)
    
    return merge(left_half, right_half)

def merge(left, right):
    result = []
    i = 0
    j = 0
    
    while(i < len(left) and j < len(right)):
        if left[i] < right[j]:
            result.append(left[i])
            i = i + 1
        else: 
            result.append(right[j])
            j = j + 1
    while(i < len(left)):
        result.append(left[i])
        i = i + 1
    while(j < len(right)):
        result.append(right[j])
        j = j + 1
    return result


a = [8,7,6,9,1,0,3,4]
sortedList = mergeSort(a)
print(sortedList)

File: Res/Sorting/MergeSort/mergeSort.dart
List<int> mergeSort(List<int> list) {
  int size = list.length;

  //Base case
  if (size <= 1) {
    return list;
  }
  int mid = size ~/ 2;
  List<int> left = list.sublist(0, mid);
  List<int> right = list.sublist(mid);

  left = mergeSort(left);
  right = mergeSort(right);

  return merge(left, right);
}

List<int> merge(List<int> left, List<int> right) {
  List<int> result = [];
  int i = 0;
  int j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] < right[j]) {
      result.add(left[i]);
      i++;
    } else {
      result.add(right[j]);
      j++;
    }
  }
  while (i < left.length) {
    result.add(left[i]);
    i++;
  }
  while (j < right.length) {
    result.add(right[j]);
    j++;
  }
  return result;
}

void main() {
  List<int> a = [78, 89, 56, 54, 12, 23, 21, 15, 58];
  List<int> sortedList = mergeSort(a);
  print(sortedList);
}
